### PR TITLE
[1/N] Add `hw_classification` to `functorch/test_aotdispatch.py` and refactor `TestPythonKey`

### PR DESCRIPTION
**Pull request by pbielak**
- Refactor `TestPythonKey` into non-parametrized class
- Add `hw_classification` to `TestAOTExport`
- Add `hw_classification` to `TestAOTDispatch`
- Add `hw_classification` to `TestEagerFusionOpInfo`
- Add `hw_classification` to `TestEagerFusionModuleInfo`

Only classes that are already device-agnostic are labelled here. The remaining ones still contain device-dependent tests at this point in the stack, so they are labelled in the PR that extracts those tests:
- `TestAOTAutograd` and its `TestAOTAutogradWithDynamo` / `TestAOTAutogradWithCache` subclasses in 2/N
- `TestPartitioning` in 6/N
- `TestAOTModuleSimplified` in 7/N

---

This is a stack of PRs to refactor the `test/functorch/test_aotdispatch.py` file:
- -> https://github.com/pytorch/pytorch/pull/196441
- https://github.com/pytorch/pytorch/pull/196442
- https://github.com/pytorch/pytorch/pull/196443
- https://github.com/pytorch/pytorch/pull/196444
- https://github.com/pytorch/pytorch/pull/196446
- https://github.com/pytorch/pytorch/pull/196447
- https://github.com/pytorch/pytorch/pull/196448

### CODE REVIEW

File: test/functorch/test_aotdispatch.py [hunk 1]
```diff
@@ -6605,6 +6610,8 @@ def forward(self, primals, tangents):
 
 
 class TestPartitioning(AOTTestCase):
+    hw_classification = HardwareClassification.GENERIC
```
Comment (etaf) [inline test/functorch/test_aotdispatch.py]: This label is applied before the class is actually device-agnostic. At this commit `TestPartitioning` still contains ~20 CUDA-gated tests that are genuinely CUDA-only: `test_autocast:7151`, the `test_functionalized_rng_codegen_*` family starting at 8099, and `test_control_deps_mixed_fwd_bw_deps_e2e:10559` (uses `torch.cuda.Stream` / `torch.cuda.Event`). Same problem at `:11149` for `TestAOTModuleSimplified`, which has 6 such tests, 3 of them additionally requiring `SM80OrLater` (`test_saved_tensors_hooks_base:12061`, `_params:12211`, `_recompile:12335`).

File: test/functorch/test_aotdispatch.py [hunk 2]
```diff
@@ -12804,6 +12813,8 @@ class TestAOTAutogradWithDynamo(TestAOTAutograd):
     These are the same as TestAOTAutograd tests, but we run dynamo first to get a graph module.
     """
 
+    hw_classification = HardwareClassification.GENERIC
```
Comment (etaf) [inline test/functorch/test_aotdispatch.py]: `TestAOTAutogradWithDynamo` and `TestAOTAutogradWithCache` (`:13091`) get labeled here, but their base `TestAOTAutograd` (`:484`) is left unclassified. That has two effects:

- The subclasses re-run 10 CUDA-gated inherited tests under a `GENERIC` label -- `test_nonidempotent_amp:4322`, `test_batch_norm_amp:4341` (specifically about the cuDNN decomposition), `test_mem_leak_from_save_for_bw:3360`, and others.
- `TestAOTAutograd`'s own ~146 tests land in the `unclassified` bucket and are excluded whenever the flag is passed, even though the same tests do run via the two subclasses.

If 2/N is where the base class gets classified, could these two labels move there as well, so the file is never in a self-contradictory state?

File: test/functorch/test_aotdispatch.py [hunk 3]
```diff
@@ -12775,23 +12790,17 @@ def test_aot_autograd_module_exhaustive(self, device, dtype, training, module_in
         aot_autograd_module_failures | symbolic_aot_autograd_module_failures,
     )
     def test_aot_autograd_symbolic_module_exhaustive(
-        self, device, dtype, training, module_info
+        self, dtype, training, module_info
```
Comment (etaf) [inline test/functorch/test_aotdispatch.py]: I'd keep the `device` parameter here. The class is still registered with `instantiate_device_type_tests(..., only_for="cpu")` at `:12803`, and the sibling `test_aot_autograd_module_exhaustive:12784` still takes and forwards `device`.

This is inert today, but it reads more like a trap than a simplification: the device suffix in the generated test name comes from `cls.device_type` (`common_device_type.py:753`), not from the signature, and `_test_aot_autograd_module_helper` uses `device` for both `module_inputs_func(device=...)` and `m.to(device)`. So if `only_for` ever grows a device, this test will report as `..._xpu` / `..._cuda` while actually running on CPU. It also isn't needed for anything else in this PR, so I'd just revert this hunk.

**Review by etaf — CHANGES_REQUESTED** [review]

File: test/functorch/test_aotdispatch.py (hunk 1 above)
Comment (pbielak) [reply test/functorch/test_aotdispatch.py]: Done

File: test/functorch/test_aotdispatch.py (hunk 2 above)
Comment (pbielak) [reply test/functorch/test_aotdispatch.py]: Done

File: test/functorch/test_aotdispatch.py (hunk 3 above)
Comment (pbielak) [reply test/functorch/test_aotdispatch.py]: Done

**Review by etaf — APPROVED** [review]
